frame/democracy/src/lib.rs

```diff
@@ -0,0 +1,2084 @@
+};
+use sp_std::prelude::*;
+
+mod conviction;
+mod types;
+mod vote;
+mod vote_threshold;
+
+#[allow(clippy::all)]
+pub mod weights;
+
+pub use conviction::Conviction;
+pub use pallet::*;
+pub use types::{
+	Delegations, OriginMap, ProposalId, ReferendumInfo, ReferendumStatus, Tally, UnvoteScope,
+};
+pub use vote::{AccountVote, Vote, Voting};
+pub use vote_threshold::{Approved, VoteThreshold};
+pub use weights::WeightInfo;
+
+#[cfg(test)]
+mod tests;
+
+const DEMOCRACY_ID: LockIdentifier = *b"democrac";
+
+/// The maximum number of vetoers on a single proposal used to compute Weight.
+///
+/// NOTE: This is not enforced by any logic.
+pub const MAX_VETOERS: u32 = 100;
+
+/// A proposal index.
+pub type PropIndex = u32;
+
+/// A referendum index.
+pub type ReferendumIndex = u32;
+
+#[allow(type_alias_bounds)]
+type BalanceOf<T: Config> = T::Balance;
+
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum PreimageStatus<AccountId, Balance, BlockNumber> {
+	/// The preimage is imminently needed at the argument.
+	Missing(BlockNumber),
+	/// The preimage is available.
+	Available {
+		data: Vec<u8>,
+		provider: AccountId,
+		deposit: Balance,
+		since: BlockNumber,
+		/// None if it's not imminent.
+		expiry: Option<BlockNumber>,
+	},
+}
+
+impl<AccountId, Balance, BlockNumber> PreimageStatus<AccountId, Balance, BlockNumber> {
+	#[allow(clippy::wrong_self_convention)]
+	fn to_missing_expiry(self) -> Option<BlockNumber> {
+		match self {
+			PreimageStatus::Missing(expiry) => Some(expiry),
+			_ => None,
+		}
+	}
+}
+
+// A value placed in storage that represents the current version of the Democracy storage.
+// This value is used by the `on_runtime_upgrade` logic to determine whether we run
+// storage migration logic.
+#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, RuntimeDebug, TypeInfo)]
+enum Releases {
+	V1,
+}
+
+#[derive(PartialEq, Eq, Clone, RuntimeDebug, Encode, Decode, TypeInfo)]
+pub enum VoteOrigin<AssetId> {
+	Vote(AssetId),
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::{
+		dispatch::DispatchResultWithPostInfo,
+		pallet_prelude::*,
+		traits::EnsureOrigin,
+		weights::{DispatchClass, Pays},
+		Parameter,
+	};
+	use frame_system::{ensure_root, ensure_signed, pallet_prelude::*};
+	use sp_runtime::{
+		traits::{AtLeast32BitUnsigned, CheckedAdd, CheckedMul, CheckedSub, Zero},
+		DispatchResult,
+	};
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + Sized {
+		type Proposal: Parameter + Dispatchable<Origin = Self::Origin> + From<Call<Self>>;
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		type Balance: Default
+			+ Parameter
+			+ Codec
+			+ Copy
+			+ Ord
+			+ CheckedAdd
+			+ CheckedSub
+			+ CheckedMul
+			+ Saturating
+			+ AtLeast32BitUnsigned
+			+ Zero
+			+ TypeInfo;
+
+		type AssetId: FullCodec
```

Comment:
inherit defi config so no need to this type here

Reply:
I see only dutch_auction pallet handles it with DeFiConfig, other pallets follow the same pattern defined here. If I have time will make this change, otherwise will take it later don't want to hold the merge because of this.

Reply:
thanks. in this case it gives additional good to keep things cleaner during copy. so it can be other base config. cc @hussein-aitlahcen this can also be applied in copy paste of vesting if suits.